## Week 10 and 11 Assignment - DATASCI200 Introduction to Data Science Programming, UC Berkeley MIDS

Write code in this Jupyter Notebook to solve the following problems. Please upload this **Notebook** with your solutions to your GitHub repository and provide a link in the last question in gradescope. 

Assignment due date: 11:59PM PT the night before the Week 12 Live Session. Do **NOT** push/upload the data file. 

## Objectives

- Explore and glean insights from a real dataset using pandas
- Practice using pandas for exploratory analysis, information gathering, and discovery
- Practice cleaning data and answering questions

## General Guidelines:

- This is a **real** dataset and so it may contain errors and other pecularities to work through
- This dataset is ~218mb, which will take some time to load (and probably won't load in Google Sheets or Excel)
- If you make assumptions, annotate them in your responses
- While there is one code/markdown cell positioned after each question as a placeholder, some of your code/responses may require multiple cells
- Double-click the markdown cells that say for example **1a answer here:** to enter your written answers. If you need more cells for your written answers, make them markdown cells (rather than code cells)
- This homework assignment is not autograded because of the variety of responses one could give. 
  - Please upload this notebook to the autograder page and the TAs will manually grade it. 
  - Ensure that each cell is run and outputs your answer for ease of grading! 
  - Highly suggest to do a `restart & run all` before uploading your code to ensure everything runs and outputs correctly.
  - Answers without code (or code that runs) will be given 0 points.
- **This is meant to simulate real world data so you will have to do some external research to determine what some of the answers are!** 

## Dataset

You are to analyze campaign contributions to the 2016 U.S. presidential primary races made in California. Use the csv file located here: https://drive.google.com/file/d/1Lgg-PwXQ6TQLDowd6XyBxZw5g1NGWPjB/view?usp=sharing. You should download and save this file in the same folder as this notebook is stored.  This file originally came from the U.S. Federal Election Commission (https://www.fec.gov/).

**DO NOT PUSH THIS FILE TO YOUR GITHUB REPO!**

- Best practice is to not have DATA files in your code repo. As shown below, the default load is outside of the folder this notebook is in. If you change the folder where the file is stored please update the first cell!
- If you do accidentally push the file to your github repo - follow the directions here to fix it: https://docs.google.com/document/d/15Irgb5V5G7pKPWgAerH7FPMpKeQRunbNflaW-hR2hTA/edit?usp=sharing

Documentation for this data can be found here: https://drive.google.com/file/d/11o_SByceenv0NgNMstM-dxC1jL7I9fHL/view?usp=sharing

## Data Questions

You are working for a California state-wide election campaign. Your boss wants you to examine historic 2016 election contribution data to see what zipcodes are more supportive of fundraising for your candidate. 

Your boss asks you to filter out some of the records:
- Only use primary 2016 contribution data (more like how your race is).
- Concentrate on Bernie Sanders as a candidate (most a like your candidate)

The questions your boss wants answered is:
- Which zipcode (5-digit zipcode) had the highest count of contributions and the most dollar amount?
- What day(s) of the month do most people donate?

## Setup

Run the cell below as it will load the data into a pandas dataframe named `contrib`. Note that a custom date parser is defined to speed up loading. If Python were to guess the date format, it would take even longer to load.

In [1]:
import pandas as pd
import numpy as np
from datetime import datetime

# These commands below set some options for pandas and to have matplotlib show the charts in the notebook
pd.set_option('display.max_rows', 1000)
pd.options.display.float_format = '{:,.2f}'.format

# Define a date parser to pass to read_csv
d = lambda x: datetime.strptime(x, '%d-%b-%y')

# Load the data
# We have this defaulted to the folder OUTSIDE of your repo - please change it as needed
contrib = pd.read_csv(r"C:\Users\vw480f\Downloads\P00000001-CA.csv", index_col=False, parse_dates=['contb_receipt_dt'], date_parser=d)

# Note - for now, it is okay to ignore the warning about mixed types. 

C:\Users\vw480f\AppData\Local\Temp\ipykernel_12976\2311071592.py:14: DtypeWarning: Columns (6,11,12) have mixed types. Specify dtype option on import or set low_memory=False.
  contrib = pd.read_csv(r"C:\Users\vw480f\Downloads\P00000001-CA.csv", index_col=False, parse_dates=['contb_receipt_dt'], date_parser=d)


***
## 1. Initial Data Checks (50 points)

First we will take a preliminary look at the data to check that it was loaded correctly and contains the info we need.

The questions to answer at the end of this section:
- Do we have the correct # of columns and rows. 
- Do the records contain data for the questions we want to answer 
- What columns are important? 
- What columns can be dropped?
- What are the data problems?

**1a.** Print the *shape* of the data. Does this match the expectation? (2 points)

In [2]:
# 1a YOUR CODE HERE
contrib.shape

(1125659, 18)

- **1a answer here:** 
Yes, the shape of the data matches the expectation. 

**1b.** Print a list of column names. Are all the columns included that are in the documentation? (2 points)

In [3]:
# 1b YOUR CODE HERE
print(contrib.columns.tolist())

['cmte_id', 'cand_id', 'cand_nm', 'contbr_nm', 'contbr_city', 'contbr_st', 'contbr_zip', 'contbr_employer', 'contbr_occupation', 'contb_receipt_amt', 'contb_receipt_dt', 'receipt_desc', 'memo_cd', 'memo_text', 'form_tp', 'file_num', 'tran_id', 'election_tp']


- **1b answer here:** 
Yes, all the columns are included that are in the documentation.

**1c** Print out the first five rows of the dataset. How do the columns `cand_id`, `cand_nm` and `contbr_st` look? (3 points)

In [4]:
# 1c YOUR CODE HERE
contrib.head(5)


cmte_id    cand_id                  cand_nm          contbr_nm  \
0  C00575795  P00003392  Clinton, Hillary Rodham         AULL, ANNE   
1  C00575795  P00003392  Clinton, Hillary Rodham  CARROLL, MARYJEAN   
2  C00575795  P00003392  Clinton, Hillary Rodham   GANDARA, DESIREE   
3  C00577130  P60007168         Sanders, Bernard          LEE, ALAN   
4  C00577130  P60007168         Sanders, Bernard   LEONELLI, ODETTE   

     contbr_city contbr_st     contbr_zip            contbr_employer  \
0       LARKSPUR        CA 949,391,913.00                        NaN   
1        CAMBRIA        CA 934,284,638.00                        NaN   
2        FONTANA        CA 923,371,507.00                        NaN   
3      CAMARILLO        CA 930,111,214.00  AT&T GOVERNMENT SOLUTIONS   
4  REDONDO BEACH        CA 902,784,310.00   VERICOR ENTERPRISES INC.   

   contbr_occupation  contb_receipt_amt contb_receipt_dt receipt_desc memo_cd  \
0            RETIRED              50.00       2016-04-26          NaN       X   
1            RETIRED             200.00       2016-04-20          NaN       X   
2            RETIRED               5.00       2016-04-02          NaN       X   
3  SOFTWARE ENGINEER              40.00       2016-03-04          NaN     NaN   
4         PHARMACIST              35.00       2016-03-05          NaN     NaN   

                             memo_text form_tp  file_num      tran_id  \
0               * HILLARY VICTORY FUND    SA18   1091718     C4768722   
1               * HILLARY VICTORY FUND    SA18   1091718     C4747242   
2               * HILLARY VICTORY FUND    SA18   1091718     C4666603   
3  * EARMARKED CONTRIBUTION: SEE BELOW   SA17A   1077404  VPF7BKWA097   
4  * EARMARKED CONTRIBUTION: SEE BELOW   SA17A   1077404  VPF7BKX3MB3   

  election_tp  
0       P2016  
1       P2016  
2       P2016  
3       P2016  
4       P2016

- **1c answer here:**

cand_id is candidate ID and starts with the letter P for both candidates. 
cand_nm is candidate name and has candidate full name with the last name separated by a comma from the first name. 
contbr_st is contributor State.

**1d.** Print out the values for the column `election_tp`. In your own words, based on the documentation, what information does the `election_tp` variable contain? Do the values in the column match the documentation? (3 points)

In [5]:
# 1d YOUR CODE HERE
print(contrib.election_tp)

contrib['election_tp'].unique()


0          P2016
1          P2016
2          P2016
3          P2016
4          P2016
           ...  
1125654    P2016
1125655    P2016
1125656    P2016
1125657    P2016
1125658    P2016
Name: election_tp, Length: 1125659, dtype: object


array(['P2016', 'G2016', nan, 'P2020'], dtype=object)

- **1d answer here:** 
election_tp is election type and this column has 2016 and 2020 primary election data, 2016 general election data, and also null values. 

**1e.** Print out the datatypes for all of the columns. What are the datatypes for the `contbr_zip`, `contb_receipt_amt`, `contb_receipt_dt`? (5 points)

In [6]:
# 1e YOUR CODE HERE
contrib.dtypes


cmte_id                      object
cand_id                      object
cand_nm                      object
contbr_nm                    object
contbr_city                  object
contbr_st                    object
contbr_zip                   object
contbr_employer              object
contbr_occupation            object
contb_receipt_amt           float64
contb_receipt_dt     datetime64[ns]
receipt_desc                 object
memo_cd                      object
memo_text                    object
form_tp                      object
file_num                      int64
tran_id                      object
election_tp                  object
dtype: object

- **1e answer here:** 

contbr_zip is object data type. 
contb_receipt_amt is float64 data type
contb_receipt_dt is datetime64[ns] data type

**1f.** What columns have the most non-nulls?  Would you recommend to drop any columns based on the number of nulls? (5 points)

In [7]:
# 1f YOUR CODE HERE
contrib.isnull().sum()


cmte_id                    0
cand_id                    0
cand_nm                    0
contbr_nm                  0
contbr_city               26
contbr_st                  0
contbr_zip                95
contbr_employer       157902
contbr_occupation      10399
contb_receipt_amt          0
contb_receipt_dt           0
receipt_desc         1110614
memo_cd               981391
memo_text             624511
form_tp                    0
file_num                   0
tran_id                    0
election_tp             1425
dtype: int64

- **1f answer here:** 
cmte_id, cand_id, cand_nm, contbr_nm, contbr_st, contb_receipt_amt ,contb_receipt_dt, form_tp, file_num, and tran_id columns have the most non-nulls. These columns have no nulls. 
I recommend dropping receipt_desc and memo_cd columns. 
memo_cd columns contain null values or the letter X which provide no useful information. 
receipt_desc also provides no important information and also have 1110614 null values. 

**1g.** A column we know that we want to use is the cand_nm column.  From the documentation each candidate is a unique candidate id also. Check data quality of `cand_id` column to see if it matches `cand_nm` column. Specifically check to ensure our targetted candidate 'Bernard Sanders' always has the same cand_id throughout. Any issues with `cand_nm` matching `cand_id`? (5 points)

In [6]:
# 1g YOUR CODE HERE

# Filtering rows for the candidate 'Bernard Sanders'
sanders_data = contrib[contrib['cand_nm'] == 'Sanders, Bernard']

# Checking unique values in the 'cand_id' column for Bernie Sanders
unique_cand_ids = sanders_data['cand_id'].unique()

print("Unique Candidate ID for Bernard Sanders:", unique_cand_ids)



# Filtering rows for the candidate 'Hillary Rodham Clinton'
clinton_data = contrib[contrib['cand_nm'] == 'Clinton, Hillary Rodham']

# Checking unique values in the 'cand_id' column for Hillary Clinton
unique_cand_ids_clinton = clinton_data['cand_id'].unique()

print("Unique Candidate ID for Hillary Clinton:", unique_cand_ids_clinton)



Unique Candidate ID for Bernard Sanders: ['P60007168']
Unique Candidate ID for Hillary Clinton: ['P00003392']


- **1g answer here:** 
There is no issue with cand_nm matching cand_id. Verified this information for Bernard Sanders and Hilary Clinton. 

**1h.** Another area to check is to make sure all of the records are from California. Check the `contbr_st` column - are there any records outside of California based on `contbr_st`? (5 points)

In [7]:
# 1h YOUR CODE HERE
unique_contbr_st=contrib['contbr_st'].unique()
print(unique_contbr_st)

['CA']


- **1h answer here:** 
There is no record outside of California as there is only one unique value for the column contbr_st, which is CA. 

**1i.** The next column to check for the analysis is the `tran_id` column. This column could be the primary key so look for duplicates. How many duplicate entries are there? Any pattern for why are there duplicate entries? (5 points)

In [8]:
# 1i YOUR CODE HERE

print(contrib['tran_id'].duplicated(keep=False).sum())

duplicate_tran_ids = contrib[contrib.duplicated(subset=['tran_id'], keep=False)]

#if not duplicate_tran_ids.empty:
    #print("Rows with Duplicate tran_id:")
    #print(duplicate_tran_ids)
#else:
    #print("No Duplicate tran_id found.")
    

ids = contrib["tran_id"]
contrib[ids.isin(ids[ids.duplicated()])].sort_values("tran_id")

6873


cmte_id    cand_id                    cand_nm             contbr_nm  \
42386  C00575449  P40003576                 Paul, Rand  PARKS, PATRICIA MRS.   
42835  C00575449  P40003576                 Paul, Rand  PARKS, PATRICIA MRS.   
43573  C00575449  P40003576                 Paul, Rand  PARKS, PATRICIA MRS.   
42428  C00575449  P40003576                 Paul, Rand       BREED, PAUL MR.   
42820  C00575449  P40003576                 Paul, Rand       BREED, PAUL MR.   
...          ...        ...                        ...                   ...   
15532  C00574624  P60006111  Cruz, Rafael Edward 'Ted'         BARTELS, DAVE   
36137  C00458844  P60006723               Rubio, Marco          RUBY, HOWARD   
37477  C00573519  P60005915        Carson, Benjamin S.        SAMPSON, RALPH   
36139  C00458844  P60006723               Rubio, Marco           RUSS, SUNNY   
37470  C00573519  P60005915        Carson, Benjamin S.       LEACH, FLORENCE   

         contbr_city contbr_st contbr_zip contbr_employer contbr_occupation  \
42386        HUGHSON        CA  95,326.00            SELF         HOUSEWIFE   
42835        HUGHSON        CA  95,326.00            SELF         HOUSEWIFE   
43573        HUGHSON        CA  95,326.00            SELF         HOUSEWIFE   
42428   SOLANA BEACH        CA  92,075.00      NET BURNER          ENGINEER   
42820   SOLANA BEACH        CA  92,075.00      NET BURNER          ENGINEER   
...              ...       ...        ...             ...               ...   
15532   LAGUNA HILLS        CA  92,653.00             NaN               NaN   
36137    LOS ANGELES        CA  90,064.00             NaN               NaN   
37477       CARLSBAD        CA  92,008.00             NaN               NaN   
36139  BEVERLY HILLS        CA  90,210.00             NaN               NaN   
37470         AUBURN        CA  95,602.00             NaN               NaN   

       contb_receipt_amt contb_receipt_dt receipt_desc memo_cd memo_text  \
42386             200.00       2015-11-28          NaN       X       NaN   
42835             200.00       2015-11-28          NaN       X       NaN   
43573             200.00       2015-11-28          NaN       X       NaN   
42428            -500.00       2015-10-28          NaN       X       NaN   
42820            -500.00       2015-10-28          NaN       X       NaN   
...                  ...              ...          ...     ...       ...   
15532          -2,700.00       2016-06-30       Refund     NaN       NaN   
36137          -1,800.00       2016-05-13       Refund     NaN       NaN   
37477            -100.00       2016-05-19       Refund     NaN       NaN   
36139            -200.00       2016-05-13       Refund     NaN       NaN   
37470             -50.00       2016-05-19       Refund     NaN       NaN   

      form_tp  file_num               tran_id election_tp  
42386   SA17A   1057796  A26C35A6066754130B99       P2016  
42835   SA17A   1057799  A26C35A6066754130B99       P2016  
43573   SA17A   1057798  A26C35A6066754130B99       P2016  
42428   SA17A   1057796  A340DF85B7F884133A20       P2016  
42820   SA17A   1057799  A340DF85B7F884133A20       P2016  
...       ...       ...                   ...         ...  
15532   SB28A   1096256           SB28A.90929         NaN  
36137   SB28A   1079053            SB28A.9119       G2016  
37477   SB28A   1079198            SB28A.9119         NaN  
36139   SB28A   1079053            SB28A.9127       G2016  
37470   SB28A   1079198            SB28A.9127         NaN  

[6873 rows x 18 columns]

- **1i answer here:** 
There are 6873 duplicate transaction IDs. There is no discernable pattern. Sometimes, there are refunds with same transaction ID, sometimes there is same file number for the same transation ID. 

**1j.** Another column to check is the `contb_receipt_amt` that shows the donation amounts. How many negative donations are included? What do negative donations mean? Please show at least pull a few rows to look at the records with negative donations. Do these records match with the expectation of why a negative donation would happen? (5 points)

In [9]:
# 1j YOUR CODE HERE
neg=contrib['contb_receipt_amt']
neg_count=neg.where(neg<0).count()
print(neg_count)

print("negative donations:")
contrib.loc[contrib['contb_receipt_amt'] < 0].head(5)


11896
negative donations:


cmte_id    cand_id                    cand_nm           contbr_nm  \
19   C00574624  P60006111  Cruz, Rafael Edward 'Ted'    JOLLIFF, RICHARD   
23   C00574624  P60006111  Cruz, Rafael Edward 'Ted'    JOLLIFF, RICHARD   
81   C00574624  P60006111  Cruz, Rafael Edward 'Ted'    JOLLIFF, RICHARD   
190  C00574624  P60006111  Cruz, Rafael Edward 'Ted'    NOWELL, DIANA L.   
213  C00574624  P60006111  Cruz, Rafael Edward 'Ted'  LICHTY, ANDREW MR.   

                contbr_city contbr_st     contbr_zip  \
19                    CHICO        CA 959,289,507.00   
23                    CHICO        CA 959,289,507.00   
81                    CHICO        CA 959,289,507.00   
190  RANCHO SANTA MARGARITA        CA 926,884,928.00   
213               SAN DIEGO        CA 921,096,720.00   

                        contbr_employer     contbr_occupation  \
19                        SELF EMPLOYED               RANCHER   
23                        SELF EMPLOYED               RANCHER   
81                        SELF EMPLOYED               RANCHER   
190  CAPISTRAND UNIFIED SCHOOL DISTRICT  LIBRARIAN TECHNICIAN   
213                       SELF EMPLOYED           REAL ESTATE   

     contb_receipt_amt contb_receipt_dt              receipt_desc memo_cd  \
19              -25.00       2016-04-29  REDESIGNATION TO GENERAL       X   
23             -150.00       2016-04-29  REDESIGNATION TO GENERAL       X   
81              -60.00       2016-04-14  REDESIGNATION TO GENERAL       X   
190            -100.00       2016-04-11  REDESIGNATION TO GENERAL       X   
213             -25.00       2016-04-30  REDESIGNATION TO GENERAL       X   

                    memo_text form_tp  file_num         tran_id election_tp  
19   REDESIGNATION TO GENERAL   SA17A   1077664  SA17A.1826482B       P2016  
23   REDESIGNATION TO GENERAL   SA17A   1077664  SA17A.1826483B       P2016  
81   REDESIGNATION TO GENERAL   SA17A   1077664   SA17A.1827494       P2016  
190  REDESIGNATION TO GENERAL   SA17A   1077664  SA17A.1639830B       P2016  
213  REDESIGNATION TO GENERAL   SA17A   1077664  SA17A.1826888B       P2016

- **1j answer here:**
There are 11896 negative donations. Negative donations are refunds. Yes, the records match with expectation. The "Redesignation to General" indicates that a contribution, originally designated for a specific purpose (primary in this case), has been reclassified or redirected to general election (which in other words means a refund or a negative amount). 

**1k.** One more column to look at is the date of donation column. Are there any dates outside of the primary period (defined as 1 Jan 2014 to 7 June 2016)? Are the dates well-formatted for our analysis? (5 points)

In [10]:
# 1k YOUR CODE HERE

contrib['contb_receipt_dt'] = pd.to_datetime(contrib['contb_receipt_dt'])

# Display rows with dates outside the primary period
outside_primary_period = contrib[(contrib['contb_receipt_dt'] < '2014-01-01') | (contrib['contb_receipt_dt'] > '2016-06-07')]

# Show the selected rows
#print("Rows with Dates Outside the Primary Period:")
print(outside_primary_period)

# Check if there are any missing or poorly formatted dates
missing_invalid_dates = contrib[contrib['contb_receipt_dt'].isnull()]

# Show the rows with missing or poorly formatted dates
print("\nRows with Missing Dates:")
print(missing_invalid_dates)


           cmte_id    cand_id                    cand_nm           contbr_nm  \
9932     C00458844  P60006723               Rubio, Marco   WHEELER, MARY MS.   
9994     C00458844  P60006723               Rubio, Marco   WHEELER, MARY MS.   
14673    C00574624  P60006111  Cruz, Rafael Edward 'Ted'     RANDALL, DICK J   
14682    C00605568  P20002671              Johnson, Gary           BOAL, ROB   
14697    C00605568  P20002671              Johnson, Gary         LEE, JASCHA   
...            ...        ...                        ...                 ...   
1123992  C00575795  P00003392    Clinton, Hillary Rodham  RANDALL, NANCY ENA   
1123993  C00575795  P00003392    Clinton, Hillary Rodham       WRIGHT, LORNA   
1123994  C00575795  P00003392    Clinton, Hillary Rodham      BALES, REBECCA   
1123995  C00575795  P00003392    Clinton, Hillary Rodham   BELLINGER, MARTHA   
1123996  C00580100  P80001571           Trump, Donald J.  DELLA DONNA, KADEE   

         contbr_city contbr_st     cont

- **1k answer here:**
Yes, there are dates outside of the primary period. The dates are well-formulated. 

**1l.** Finally, answer the initial questions in the cells below (5 points)

**1l.1** Do we have the correct # of columns and rows.

- **1l.1 answer here:**
Yes, we have the correct # of columns and rows as expected. 

**1l.2** Do the records contain data for the questions we want to answer?

- **1l.2 answer here:**
Yes the records contain data for the questions we want to answer.

**1l.3** What columns are important?

- **1l.3 answer here:** 
contbr_zip, contb_receipt_amt and,contb_receipt_dt are important columns. 

**1l.4** What columns can be dropped?

- **1l.4 answer here:** 
cmte_id, contbr_employer, contbr_occupation,receipt_desc, memo_cd, memo_text, form_tp, file_num, tran_id, election_tp, contbr_st, contbr_nm can be dropped. 
contbr_city and cand_nm can also be dropped, but they can be useful in further analysis. For example, cand_nm will be needed to find out candidates name based on cand_id. 


**1l.5** What are the data problems?

- **1l.5 answer here:**
The accuracy of the data cannot be verified. It is also not know if any positive donations have been entered as negative amount by mistake. 

**1l.6** List any assumptions so far:

- **1l.6 answer here:**
Assumptions:
a. All the refunds have been recorded as negative amount in contb_receipt_amt column. 
b. Data integrity: All entries are accurate, complete, and consistent. 

***
## 2. Data filtering and data quality fixes (30 points)

Now that we have a basic understanding of the data, let's filter out the records we don't need and fix the data.

**2a.** From the dataset filter out (remove) any election_tp not in the primary election. Print/show the shape of the dataframe after the filtering is complete. (5 points)

In [11]:
# 2a YOUR CODE HERE

primary_contributions_data = contrib[contrib['election_tp'] == 'P2016']

# Print the shape of the DataFrame after filtering
print("Shape of DataFrame after Filtering:", primary_contributions_data.shape)


Shape of DataFrame after Filtering: (810481, 18)


**2b.** From the dataset filter out (remove) any candidate that is not Bernie Sanders. Print/show the shape of the dataframe after the filtering is complete. (5 points)

In [12]:
# 2b YOUR CODE HERE
primary_contributions_data_Bernie = primary_contributions_data[primary_contributions_data['cand_id'] == 'P60007168']

# Print the shape of the DataFrame after filtering
print("Shape of DataFrame after Filtering:", primary_contributions_data_Bernie.shape)


Shape of DataFrame after Filtering: (407171, 18)


**2c.** The `contbr_zip` column is not formatted well for our analysis. Make a new zipcode column that is the five-digit zipcodes. Filter out any records outside of California based on the zipcode. Print/show the shape of the dataframe after the filtering is complete. (10 points).

- You will have to research what the valid 5-digit zipcodes for California are!

In [13]:
# 2c YOUR CODE HERE

""" California zipcode ranges between 90001 and 961062"""

# Dropping rows with missing values in the 'contbr_zip' column
contrib_cleaned = primary_contributions_data_Bernie.dropna(subset=['contbr_zip'])


# Extracting the first five characters from 'contbr_zip' to get five-digit zip codes
contrib_cleaned['contbr_zip'] = contrib_cleaned['contbr_zip'].astype(str)
contrib_cleaned['zipcode'] = contrib_cleaned['contbr_zip'].str[:5]

# Converting 'zipcode' column to numeric, replacing non-numeric values with NaN
contrib_cleaned['zipcode'] = pd.to_numeric(contrib_cleaned['zipcode'], errors='coerce')


# Filtering out records for California zipcode 
california_contributions = contrib_cleaned[
    (contrib_cleaned['zipcode'] >= 90001) &
    (contrib_cleaned['zipcode'] <= 96162)]

print(california_contributions)
# Printing the shape of the DataFrame after filtering
print("Shape of DataFrame after Filtering for California Zip Codes:", california_contributions.shape)

# Checking for NaN values in 'zipcode'
print("NaN values in 'zipcode':", california_contributions['zipcode'].isna().any())




           cmte_id    cand_id           cand_nm            contbr_nm  \
3        C00577130  P60007168  Sanders, Bernard            LEE, ALAN   
4        C00577130  P60007168  Sanders, Bernard     LEONELLI, ODETTE   
5        C00577130  P60007168  Sanders, Bernard     LEONELLI, ODETTE   
6        C00577130  P60007168  Sanders, Bernard       LEOPARD, PATTI   
8        C00577130  P60007168  Sanders, Bernard         LEPKE, KELLY   
...            ...        ...               ...                  ...   
1121696  C00577130  P60007168  Sanders, Bernard       SIMONE, ARLENE   
1121698  C00577130  P60007168  Sanders, Bernard       SIMONE, ARLENE   
1121703  C00577130  P60007168  Sanders, Bernard  ROBINSON, KATHARINE   
1121704  C00577130  P60007168  Sanders, Bernard  ROBINSON, KATHARINE   
1121705  C00577130  P60007168  Sanders, Bernard  ROBINSON, KATHARINE   

           contbr_city contbr_st   contbr_zip            contbr_employer  \
3            CAMARILLO        CA  930111214.0  AT&T GOVERNM

**2d.** The receipt amount column has negative donations. After talking with your team, a decision was made that the best course of action is to remove these negative values so that the donation count and amount is more accurate. Print/show the shape of the dataframe after the filtering is complete. (5 points)

In [15]:
# 2d YOUR CODE HERE

# Filter out rows with negative values in 'contb_receipt_amt'
contrib_filtered = contrib_cleaned[contrib_cleaned['contb_receipt_amt'] >= 0]

# Print the shape of the DataFrame after filtering
print("Shape of DataFrame after removing negative donations:", contrib_filtered.shape)


Shape of DataFrame after removing negative donations: (404083, 19)


**2e.** From the dataset drop any columns that won't be used in the analysis. Print/show the shape of the dataframe after the dropping is complete. What columns did you drop and why? (5 points)

In [16]:
# 2e YOUR CODE HERE
analysis_data=contrib_filtered.drop(columns=['cmte_id','cand_id','contbr_nm','contbr_st','contbr_employer','contbr_occupation',
                              'receipt_desc','memo_cd','memo_text','form_tp','file_num','tran_id','election_tp','contbr_city',
                              'cand_nm','contbr_zip'])

print(analysis_data)
analysis_data.shape

         contb_receipt_amt contb_receipt_dt   zipcode
3                    40.00       2016-03-04 93,011.00
4                    35.00       2016-03-05 90,278.00
5                   100.00       2016-03-06 90,278.00
6                    25.00       2016-03-04 92,084.00
8                    10.00       2016-03-05 92,683.00
...                    ...              ...       ...
1121696              50.00       2016-04-23 91,436.00
1121698             100.00       2016-04-27 91,436.00
1121703              15.00       2016-04-15 92,831.00
1121704              15.00       2016-04-23 92,831.00
1121705              15.00       2016-04-25 92,831.00

[404083 rows x 3 columns]


(404083, 3)

- **2e answer here:**
Shape is [404083 rows x 3 columns]. 
I dropped all columns except contirb_receipt_amt, contb_receipt_dr, and zipcode columns. This is because all the other columns are not needed for the analysis. 

zipcode and contb_receipt_amt column are needed to find zipcode with the highest count of contributions and also to find the most dollar amount. 

contb_receipt_dt column is needed to find the days of the month with most number of contributions. 

Since the data is already filtered for Bernie Sanders and Primary election and california donations, we do not need the columns containing those information. The rest of the columns that have been removed are also not needed in the analysis. 

**2f.** List any assumptions that you made up to this point:

- **2f answer here:**
Assumptions:
a. Data are accurate and there is no mixing of Sanders and Hilary data.
b. All refunds and redesignation amounts have been correctly logged as negative amount. All other donations are recorded as 
   positive amount. 


***
## 3. Answering the questions (20 points)

Now that the data is cleaned and filterd - let's answer the two questions from your boss!

**3a.** Which zipcode had the highest count of contributions and the most dollar amount? (10 points)

In [17]:
# 3a YOUR CODE HERE
# Grouping by 'zipcode' and calculating the count and sum for each zipcode
zipcode_stats = analysis_data.groupby('zipcode').agg({'contb_receipt_amt': ['count', 'sum']})

# Rename the columns for clarity
zipcode_stats.columns = ['contribution_count', 'total_dollar_amount']

# Find the zipcode with the highest count of contributions
highest_count_zipcode = zipcode_stats['contribution_count'].idxmax()

# Find the zipcode with the highest total dollar amount
highest_amount_zipcode = zipcode_stats['total_dollar_amount'].idxmax()

print(f"The zipcode with the highest count of contributions: {highest_count_zipcode}")
print(f"The zipcode with the most dollar amount: {highest_amount_zipcode}")


The zipcode with the highest count of contributions: 94110.0
The zipcode with the most dollar amount: 94110.0


- **3a answer here:** 

**3b.** What day(s) of the month do most people donate? (10 points)

In [18]:
# 3b YOUR CODE HERE

# Converting 'contb_receipt_dt' to datetime format
analysis_data['contb_receipt_dt'] = pd.to_datetime(analysis_data['contb_receipt_dt'])

# Extracting the day of the month
analysis_data['day_of_month'] = analysis_data['contb_receipt_dt'].dt.day

# Grouping by day of the month and calculating the count of donations
day_stats = analysis_data.groupby('day_of_month')['contb_receipt_amt'].count()

# Finding the day(s) with the highest count of donations
most_donated_days = day_stats.idxmax()

print(f"The day(s) of the month with the most donations: {most_donated_days}")



The day(s) of the month with the most donations: 29


- **3b answer here:** 

## If you have feedback for this homework, please submit it using the link below:

http://goo.gl/forms/74yCiQTf6k